oci/layer/xattr.go

```diff
@@ -0,0 +1,214 @@
+	// the xattr will not be included in the tar archive.
+	//
+	// Note that ToTar is not exclusively called on tar archive entries from
+	// layer archives. When doing UnpackEntry we need to restore the metadata
+	// of the parent directory of any path we extract into (to ensure that
+	// {a,c,m}times are correct), and an implementation detail of UnpackEntry
+	// is that this involves converting the on-disk information of the parent
+	// directory into a tar.Header and then back into the on-disk
+	// representation (to ensure that extracting entry metadata is identical to
+	// restoring parent directory metadata). This means that ToTar will be
+	// routinely called on directories that may have "special" xattrs.
+	//
+	// [MaskedOnDisk] must only return true for a given xattr and
+	// [WhiteoutMode] *if and only if*, ToTar with the same arguments returns
+	// nil. Otherwise, when the metadata is re-applied with on-disk data you
+	// will end up losing on-disk xattrs or adding new nonsense xattrs.
+	ToTar(onDiskMode WhiteoutMode, xattr string) (newName *string)
+}
+
+// forbiddenXattrFilter is a dummy filter that will block all xattrs that are
+// associated with it.
+type forbiddenXattrFilter struct{}
+
+var _ xattrFilter = forbiddenXattrFilter{}
+
+func (_ forbiddenXattrFilter) MaskedOnDisk(_ WhiteoutMode, _ string) bool { return true }
+func (_ forbiddenXattrFilter) ToDisk(_ WhiteoutMode, _ string) *string    { return nil }
+func (_ forbiddenXattrFilter) ToTar(_ WhiteoutMode, _ string) *string     { return nil }
+
+// overlayXattrFilter is a filter for all trusted.overlay.* xattrs which will
+// escape the xattrs on unpack and unescape them when
+type overlayXattrFilter struct{}
+
+var _ xattrFilter = overlayXattrFilter{}
+
+func (_ overlayXattrFilter) MaskedOnDisk(woMode WhiteoutMode, xattr string) bool {
+	// Only trusted.overlay.* top-level xattrs are masked in overlayfs mode.
+	// Escaped xattrs and xattrs in regular mode are allowed.
+	return woMode == OverlayFSWhiteout &&
+		doesXattrMatch(xattr, "trusted.overlay.") &&
+		!doesXattrMatch(xattr, "trusted.overlay.overlay.")
+}
+
+func (_ overlayXattrFilter) ToDisk(woMode WhiteoutMode, xattr string) *string {
+	// If we are not in overlayfs extraction mode, we can just extract the
+	// xattr as normal.
+	if woMode != OverlayFSWhiteout {
+		return &xattr
+	}
+
+	subXattr, found := strings.CutPrefix(xattr, "trusted.overlay.")
+	if !found {
+		// We should never hit this case in practice.
+		// TODO: Should we just panic here?
+		log.Warnf("[internal error] overlayfs xattr filter called for non-overlayfs xattr %q", xattr)
+		return &xattr
+	}
+	escaped := "trusted.overlay.overlay." + subXattr
+	return &escaped
+}
+
+func (_ overlayXattrFilter) ToTar(woMode WhiteoutMode, xattr string) *string {
+	// If we are not in overlayfs extraction mode, we can just pack the xattr
+	// as normal. However, this is a bit of a strange situation so we should
+	// output a warning.
```

Comment:
Hm? I guess it's not that strange, isn't it the default?

Reply:
This comment probably should be removed now that we no longer log anything here, but this is kind of unusual -- you wouldn't expect to see `trusted.overlay.foo` xattrs in a regular image or a regular rootfs. I would assume it's a configuration error honestly (you forgot to specify the on-disk format is overlayfs).

Reply:
What if the filesystem has overlay whiteouts applied though? Like in stacker's case, if someone does a delete which generates an overlay whiteout xattr, but you're converting back to tar, won't this code get called?

But in any case, I agree that just deleting the comment is fine.

Reply:
No, in that case it gets handled in `oci/layer/generate.go` (where `AddWhiteout` gets called instead of `Add`), same as before.